workflow/vm/vm.go

```diff
@@ -0,0 +1,164 @@
+)
+
+// VM is a virtual machine that runs workflows.
+type VM struct {
+	// ExecuctionRequests keeps execution requests, caller should read this chan
+	// and perform executions.
+	ExecuctionRequests chan *Execution
+
+	m               sync.Mutex // protects following.
+	workflows       []workflow.Workflow
+	executionChains []*executionChain
+	closing         bool
+}
+
+// New creates a new Workflow VM.
+func New() *VM {
+	return &VM{
+		ExecuctionRequests: make(chan *Execution),
+	}
+}
+
+// Process processes an event by creating execution requests for matching workflows.
+func (v *VM) Process(e Event) {
+	v.m.Lock()
+	defer v.m.Unlock()
+	if v.closing {
+		return
+	}
+	if e.Secret != "" {
+		// if secret is non-empty, event is actually a result of previous execution request.
+		// generate the next execution request on that chain by using data from this result.
+		for i, c := range v.executionChains {
+			ok, execReq := c.Next(e)
+			// check if we're at the end of the chain, if so, workflow's lifecycle for this
+			// chain has completed and it can be removed.
+			if c.IsDone() {
+				v.executionChains = append(v.executionChains[:i], v.executionChains[i+1:]...)
+			}
+			if ok {
+				v.ExecuctionRequests <- execReq
+				break
+			}
+		}
+	}
+	// event can also be a start of some new lifecycles for many workflows.
+	// create a full trigger spec for the event and find out the matching workflows.
+	t := e.createTrigger()
+	for _, w := range v.workflows {
+		// check if workflow matches with trigger created from event.
+		if ok := w.Match(t); ok {
+			// create a new chain, register it and create task execution request for
+			// its first task.
+			c := newChain(w)
+			ok, execReq := c.Next(e)
+			// chain might have just one task, if so, don't bother to add it chain list.
+			if !c.IsDone() {
+				v.executionChains = append(v.executionChains, c)
+			}
+			// check against empty workflow Tasks.
+			if ok {
+				v.ExecuctionRequests <- execReq
+			}
+		}
+	}
+}
+
+// Add adds a workflow to VM.
+func (v *VM) Add(w workflow.Workflow) {
+	v.m.Lock()
+	defer v.m.Unlock()
+	if v.closing {
+		return
+	}
+	v.workflows = append(v.workflows, w)
+}
+
+// Remove removes workflow and all of its execution chains from VM.
+func (v *VM) Remove(hash string) {
+	v.m.Lock()
+	defer v.m.Unlock()
+	if v.closing {
+		return
+	}
+	for i, w := range v.workflows {
+		if w.Hash == hash {
+			v.workflows = append(v.workflows[:i], v.workflows[i+1:]...)
+			for ii := len(v.executionChains) - 1; i >= 0; i-- {
+				if v.executionChains[ii].w.Hash == hash {
+					v.executionChains = append(v.executionChains[:ii], v.executionChains[ii+1:]...)
+				}
+			}
+			return
+		}
+	}
+}
+
+// Shutdown immediately ends vm and closes ExecutionRequests channel.
+func (v *VM) Shutdown() {
+	v.m.Lock()
+	defer v.m.Unlock()
+	v.closing = true
+	close(v.ExecuctionRequests)
+}
+
+// Event represents a new event/result input to the workflow VM.
+type Event struct {
```

Comment:
Why Execution is redefined here? Shound't we use event.Event?

Reply:
https://github.com/mesg-foundation/engine/pull/1048#discussion_r293338121

also, event here is combination of Event & Result. `event.Event` just represents the Event itself. In future, we'll be merging Result to Event as discussed with @antho1404.